src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1896,6 +1896,16 @@ private IEnumerable<TypeInferenceInfo> InferTypeForReturnStatement(ReturnStateme
                     }
                 }
 
+                // If we are inside a local function then use the return type of the local function
+                var localFunctionStatement = returnStatement.GetAncestor<LocalFunctionStatementSyntax>();
```

Comment:
I don't think this is correct. If you have a local function inside a lambda, the previous checks will win and this code will never reach. What we really want is the first ancestor that is *either* a lambda or a local function.

Reply:
I'll create and add some tests, which should fail if your analysis is correct (good thinking either way!). I'll get to it later today.